ksqldb-engine/src/main/java/io/confluent/ksql/util/BinPackedStreamsMetadata.java

```diff
@@ -0,0 +1,199 @@
+    queryMap = new HashMap<>();
+    metadata = new HashMap<>();
+    listeners = new HashMap<>();
+  }
+
+  public void addQuery(final QueryErrorClassifier errorClassifier,
+                       final Map<String, Object> streamsProperties,
+                       final BinPackedPersistentQueryMetadataImpl binPackedPersistentQueryMetadata,
+                       final QueryId queryId) {
+    this.errorClassifier = errorClassifier;
+    this.streamsProperties =
+        ImmutableMap.copyOf(
+            Objects.requireNonNull(streamsProperties, "streamsProperties"));
+    this.metadata.put(queryId, binPackedPersistentQueryMetadata);
+    this.listeners.put(queryId, binPackedPersistentQueryMetadata.getListener());
+  }
+
+  protected StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse uncaughtHandler(
+      final Throwable e
+                                                                                         ) {
+    QueryError.Type errorType = QueryError.Type.UNKNOWN;
+    try {
+      errorType = errorClassifier.classify(e);
+    } catch (final Exception classificationException) {
+      LOG.error("Error classifying unhandled exception", classificationException);
+    } finally {
+      // If error classification throws then we consider the error to be an UNKNOWN error.
+      // We notify listeners and add the error to the errors queue in the finally block to ensure
+      // all listeners and consumers of the error queue (e.g. the API) can see the error. Similarly,
+      // log in finally block to make sure that if there's ever an error in the classification
+      // we still get this in our logs.
+      final QueryError queryError =
+          new QueryError(
+              System.currentTimeMillis(),
+              Throwables.getStackTraceAsString(e),
+              errorType
+          );
+      for (Map.Entry<QueryId, QueryMetadata.Listener> listener: listeners.entrySet()) {
+        listener.getValue().onError(metadata.get(listener.getKey()), queryError);
+      }
+      queryErrors.add(queryError);
+      LOG.error(
+          "Unhandled exception caught in streams thread {}. ({})",
+          Thread.currentThread().getName(),
+          errorType,
+          e
+      );
+    }
+    return StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse.REPLACE_THREAD;
+  }
+
+  public KafkaStreams getKafkaStreams() {
+    return kafkaStreams;
+  }
+
+  public KafkaStreams.State state() {
+    return kafkaStreams.state();
+  }
+
+  public Collection<StreamsMetadata> allMetadata() {
+    return kafkaStreams.allMetadata();
+  }
+
+  public Set<StreamsTaskMetadata> getTaskMetadata() {
+    return kafkaStreams.localThreadsMetadata()
+                       .stream()
+                       .flatMap(t -> t.activeTasks().stream())
+                       .map(StreamsTaskMetadata::fromStreamsTaskMetadata)
+                       .collect(Collectors.toSet());
+  }
+
+  public void addToplogy(final Topology topology) {
```

Comment:
Maybe we can just call this `start(queryId)`?

Reply:
My previous comment may be a bis unclear, I'm just wondering why we cannot just call the other

```
public void start(final QueryId queryId) {
    kafkaStreams.addTopology(queryMap.get(queryId.toString()));
  }
```

whenever we need to call this? again I'm feeling that with a single consolidated map we should just need to `start/addTopology` once, together.

Reply:
ah that might work, if you can't tell I added the consolidated map after I had designed it and haven't worked out all the ways I can use it best yet :)